sqldb/invoices.go

```diff
@@ -0,0 +1,1651 @@
+		paymentRequest []byte
+		preimage       *lntypes.Preimage
+		paymentAddr    [32]byte
+	)
+
+	hash, err := lntypes.MakeHash(row.Hash)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if row.SettledIndex.Valid {
+		settledIndex = row.SettledIndex.Int64
+	}
+
+	if row.SettledAt.Valid {
+		settledAt = row.SettledAt.Time.Local()
+	}
+
+	if row.Memo.Valid {
+		memo = []byte(row.Memo.String)
+	}
+
+	// Keysend payments will have this field empty.
+	if row.PaymentRequest.Valid {
+		paymentRequest = []byte(row.PaymentRequest.String)
+	} else {
+		paymentRequest = []byte{}
+	}
+
+	// We may not have the preimage if this a hodl invoice.
+	if row.Preimage != nil {
+		preimage = &lntypes.Preimage{}
+		copy(preimage[:], row.Preimage)
+	}
+
+	copy(paymentAddr[:], row.PaymentAddr)
+
+	var cltvDelta int32
+	if row.CltvDelta.Valid {
+		cltvDelta = row.CltvDelta.Int32
+	}
+
+	invoice := &invpkg.Invoice{
+		SettleIndex:    uint64(settledIndex),
+		SettleDate:     settledAt,
+		Memo:           memo,
+		PaymentRequest: paymentRequest,
+		CreationDate:   row.CreatedAt.Local(),
+		Terms: invpkg.ContractTerm{
+			FinalCltvDelta:  cltvDelta,
+			Expiry:          time.Duration(row.Expiry),
+			PaymentPreimage: preimage,
+			Value:           lnwire.MilliSatoshi(row.AmountMsat),
+			PaymentAddr:     paymentAddr,
+		},
+		AddIndex:    uint64(row.ID),
+		State:       invpkg.ContractState(row.State),
+		AmtPaid:     lnwire.MilliSatoshi(row.AmountPaidMsat),
+		Htlcs:       make(map[models.CircuitKey]*invpkg.InvoiceHTLC),
+		AMPState:    invpkg.AMPInvoiceState{},
+		HodlInvoice: row.IsHodl,
+	}
+
+	return &hash, invoice, nil
+}
+
+// unmarshalInvoiceHTLC converts an sqlc.InvoiceHtlc to an InvoiceHTLC.
+func unmarshalInvoiceHTLC(row sqlc.InvoiceHtlc) (invpkg.CircuitKey,
+	*invpkg.InvoiceHTLC, error) {
+
+	uint64ChanID, err := strconv.ParseUint(row.ChanID, 10, 64)
+	if err != nil {
+		return invpkg.CircuitKey{}, nil, err
+	}
+
+	chanID := lnwire.NewShortChanIDFromInt(uint64ChanID)
+
+	if row.HtlcID < 0 {
+		return invpkg.CircuitKey{}, nil, fmt.Errorf("invalid uint64 "+
+			"value: %v", row.HtlcID)
+	}
+
+	htlcID := uint64(row.HtlcID)
+
+	circuitKey := invpkg.CircuitKey{
+		ChanID: chanID,
+		HtlcID: htlcID,
+	}
+
+	htlc := &invpkg.InvoiceHTLC{
+		Amt:          lnwire.MilliSatoshi(row.AmountMsat),
+		AcceptHeight: uint32(row.AcceptHeight),
+		AcceptTime:   row.AcceptTime.Local(),
+		Expiry:       uint32(row.ExpiryHeight),
+		State:        invpkg.HtlcState(row.State),
+	}
+
+	if row.TotalMppMsat.Valid {
+		htlc.MppTotalAmt = lnwire.MilliSatoshi(row.TotalMppMsat.Int64)
+	}
+
+	if row.ResolveTime.Valid {
+		htlc.ResolveTime = row.ResolveTime.Time.Local()
+	}
+
+	return circuitKey, htlc, nil
+}
+
+// queryWithLimit is a helper method that can be used to query the database
+// using a limit and offset. The passed query function should return the number
+// of rows returned and an error if any.
+func queryWithLimit(query func(int) (int, error), limit int) error {
```

Comment:
This function is super confusing for me. How are we supposed to use it?

Reply:
I wonder why we have to throttle the db. requests on the application side rather than letting the sql backend handle this ?

Reply:
I'm not actually sure that we can just rely on the backend and ask it to return everything in one package. I'm thinking that we may be hitting certain memory limits if the returned set is exceedingly large etc.